chrome/0001-add-tests-for-sourcemaps-spec.patch

```diff
@@ -0,0 +1,2070 @@
++++ b/front_end/BUILD.gn
+@@ -106,6 +106,7 @@ group("unittests") {
+     "core/protocol_client:unittests",
+     "core/root:unittests",
+     "core/sdk:unittests",
++    "core/sdk/fixtures/sourcemaps",
+     "entrypoints/formatter_worker:unittests",
+     "entrypoints/heap_snapshot_worker:unittests",
+     "entrypoints/inspector_main:unittests",
+diff --git a/front_end/core/sdk/BUILD.gn b/front_end/core/sdk/BUILD.gn
+index 43ea1be19c..bdc88a7e04 100644
+--- a/front_end/core/sdk/BUILD.gn
++++ b/front_end/core/sdk/BUILD.gn
+@@ -163,6 +163,7 @@ ts_library("unittests") {
+     "SourceMap.test.ts",
+     "SourceMapManager.test.ts",
+     "SourceMapScopes.test.ts",
++    "SourceMapSpec.test.ts",
+     "StorageBucketsModel.test.ts",
+     "StorageKeyManager.test.ts",
+     "Target.test.ts",
+diff --git a/front_end/core/sdk/SourceMapSpec.test.ts b/front_end/core/sdk/SourceMapSpec.test.ts
+new file mode 100644
+index 0000000000..64f350d7f3
+--- /dev/null
++++ b/front_end/core/sdk/SourceMapSpec.test.ts
+@@ -0,0 +1,129 @@
++// Copyright 2024 The Chromium Authors. All rights reserved.
++// Use of this source code is governed by a BSD-style license that can be
++// found in the LICENSE file.
++
++
++/**
++ This file tests if devtools sourcemaps implementation is matching the sourcemaps spec.
++ Sourcemap Spec tests are using test data coming from: https://github.com/takikawa/source-map-tests
++
++ At the moment only basic mapping tests are implemented. Expected results:
++
++  ==== FAIL: SourceMapSpec/checks mappings for valid-mapping-null-sources.js.map
++  AssertionError: unexpected source URL: expected 'null' to equal null
++      at front_end/core/sdk/SourceMapSpec.test.ts:74:20 <- out/Default/gen/front_end/core/sdk/SourceMapSpec.test.js:25:32
++      at Array.forEach (<anonymous>)
++      at Context.<anonymous> (front_end/core/sdk/SourceMapSpec.test.ts:62:21 <- out/Default/gen/front_end/core/sdk/SourceMapSpec.test.js:21:29)
++
++  - expected
++  + actual
++
++  -[null]
++  +"null"
++  ==============================================================================
++
++  FAILED: 1 failed, 46 passed (0 skipped)
++  ERRORS DETECTED
++
++ **/ 
++
++const {assert} = chai;
++import type * as Platform from '../platform/platform.js';
++import {assertNotNullOrUndefined} from '../platform/platform.js';
++import { SourceMapV3 } from './SourceMap.js';
++import * as SDK from './sdk.js';
++import {describeWithEnvironment} from '../../testing/EnvironmentHelpers.js';
++
++interface TestSpec {
++  name: string;
++  description: string;
++  baseFile: string;
++  sourceMapFile: string;
++  sourceMapIsValid: boolean;
++  testActions?: TestAction[];
++} 
++
++interface TestAction {
++  actionType: string;
++  generatedLine: number;
++  generatedColumn: number;
++  originalSource: string;
++  originalLine: number;
++  originalColumn: number;
++  mappedName: null | string;
++}
++
++const testCases = await loadTestCasesFromFixture('source-map-spec-tests.json');
```

Comment:
Thanks! This looks great already!

I'm not sure yet how best to pull in the tests. Ideally the test cases are published either as an npm module, or we pull in the whole git repo. The reason is that we wanna keep it as simple as possible to sync the tests with the latest upstream version and introduce as few modifications downstream as possible.

I don't think we'll be able to load the test cases like this. The reason is how Karma and Mocha work: Karma lists each test file as a separate `<script module="..."></script>` tag in a generated HTML file. Once all the modules have loaded, Mocha starts running. This means we can't create test cases (`it`-blocks) asynchronously. They might be created after Mocha already thought it knows about all the test cases and starts executing the suites. Similarly nested `it`-blocks are also not allowed. Basically, Mocha requires all `describe`/`it` invocations to run synchronously, so when all the `<script module=".../.test.js">` modules are done, Mocha knows all the tests.

To get around this, we have two possibilities (in decreasing order of desirability):
  1. We use a static import to get the 'source-map-spec-tests.json'. With import assertions this should be doable now.
  1. We put all the spec test cases into a single `it` invocation. Basically `it('follows the spec', () => {`. We can load the test cases in the `before` hook for that single `it`.
  1. We do some code generation at build time. Basically as a GN step where we invoke a script that reads the JSON and spits out a JS file with generated it blocks - one for each spec test case.

Reply:
@szuend thanks for checking it out and for your comments! 
I also think that ideally the test cases will be available as an npm package and can be just pulled in this way.

What do you mean that we will not be able to load the tests like this? Do you mean your CI? Locally in the `devtools-frontend` cehckout the tests are passing when I run them with `npm run auto-unittest` (except one which fails accurately because it is expecting a different value). Is there another way I can run them to make sure we don't get into the problem you're decribing?

The biggest problem I had here was actually loading the files. I haven't found any test loading a json except the trace test files loading a gzipped json e.g. `await TraceLoader.allModels(this, 'slow-interaction-button-click.json.gz');`. And I followed the same approach used there - just fetching the files from the test server.

It looks like importing a json will not work with current setup because ts is complaining during the build. Changing ts config to `resolveJsonModule: true` will not work with current module resolution, I'm getting this error:

```
gen/front_end/generated/generated-tsconfig.json(22,5): error TS5070: Option '--resolveJsonModule' cannot be specified when 'moduleResolution' is set to 'classic'
```

That's why fetching it directly from the karma server and using a simple fetch was a solution I went for so I don't have to change those main ts settings.

Let me know what you think.

Reply:
Sorry for the late reply, I was OOO, May is notorious in Germany for a series of public holidays.

It's unfortunate that importing JSON via ES `import` is not working out of the box, but we may be able to tweak the tsconfig for `testonly` Node.js targets to make this work.

I apologise, I'm struggling a bit to properly explain the technical limitation of Mocha here, but the gist of it is: `describe` and `it` blocks need to be created synchronously at ES module instantiation time. Otherwise they may or may not run flakily. The source map tests ran for you locally now, but they may not in the future. The `TraceLoader` is indeed used to load fixtures, but it's always done in a `beforeEach` block, never to instantiate test cases (`describe`/`it` blocks) dynamically after the loading is done.

Reply:
Thanks for the suggestions! I got it now. I'll try those 2 (maybe 3) things next:
- changing tsconfig for tests to enable json imports,
- moving the loading of sourcemaps into the beforeEach block and running the tests in a single it block
- _optionally_ figure out how to generate separate tests for different source maps  with code generation (not sure yet if this would be nicer than a single if, might be a bit too much as there is a lot of test files).